recursive-methods.rb

```diff
@@ -1,36 +1,258 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
+
+# Write #1
+# factorial(n) Write a method factorial that accepts an integer parameter n and that uses recursion to compute and return the value of n factorial (also known as n!)
+# e.g. fact(4) = 4 * 3 * 2 * 1 = 24
+
 def factorial(n)
-  puts "Not implemented."
+  if n == 0
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
+end
+
+if factorial(4) == 24
+  puts "Factorial alg: Right!"
+else
+  puts "Factorial alg: BUG"
 end
 
+#time comp: O(n) where n is the size of the digit. As the integer increases, the computational time increases at the same rate.
+#space comp: O(n) where n is the size of the number. Ad the number increases, the number of times the function is called grows at the same rate.
+
+# Write #2
+# reverse(s) Write a method reverse that accepts a string as a parameter and then returns the reverse of the string by reversing all letters and all words in the string.
+# e.g. reverse("hello world") will return "dlrow olleh"
+
 def reverse(s)
-  puts "Not implemented."
+  n = s.length - 1
```

Comment:
This works! See if you can update this method to not modify the original string s, and instead return a newly created string object with characters reversed.